node_modules/oae-content/lib/api.js

```diff
@@ -1079,48 +1079,75 @@ var setContentPermissions = module.exports.setContentPermissions = function(ctx,
             return callback(err);
         }
 
-        // Check if the current user can manage the content
-        ContentAuthz.canManageContent(ctx, contentObj, function(err, canManage) {
+        // Check that the user isn't trying to assign any role higher than his own
+        AuthzAPI.resolveEffectiveRole(ctx, contentObj.id, contentObj.tenant.alias, contentObj.visibility, ContentConstants.roles.ALL_PRIORITY, function(err, effectiveRole, canInteract) {
             if (err) {
                 return callback(err);
-            } else if (!canManage) {
-                return callback({'code': 401, 'msg': 'You are not allowed to manage this piece of content'});
+            } else if (!canInteract) {
+                return callback({'code': 401, 'msg': 'You are not allowed to interact with this piece of content'});
             }
 
-            // Ensure that we're not crossing tenant boundaries or violating principal visibility settings
-            _checkNewContentPermissions(ctx, contentObj, permissionChanges, function(err, membershipAfterChanges, newMemberIds, updatedMemberIds, removedMemberIds, newMembers) {
+            ContentAuthz.canManageContent(ctx, contentObj, function(err, canManage) {
                 if (err) {
                     return callback(err);
                 }
 
-                // Apply the permission changes to the content. This will then update the content libraries as well
-                _setContentPermissions(ctx, contentObj, permissionChanges, function(err, newContentObj) {
-                    if (err) {
-                        return callback(err);
+                AuthzAPI.getDirectRoles(_.keys(permissionChanges), contentObj.id, function(err, existingRoles) {
+                    if (!canManage) {
+                        if (_.contains(permissionChanges, false)) {
+                            return callback({'code': 401, 'msg': 'Only managers can remove access'});
+                        }
+
+                        var settingHigherRole = _.find(permissionChanges, function(role, principalId) {
+                            return ContentConstants.roles.ALL_PRIORITY.indexOf(effectiveRole) < ContentConstants.roles.ALL_PRIORITY.indexOf(role);
+                        });
+                        if (settingHigherRole) {
+                            return callback({'code': 401, 'msg': 'You cannot set another user\'s role higher than your own'});
+                        }
+                        var hasHigherRole = _.find(existingRoles, function(role, principalId) {
+                            return ContentConstants.roles.ALL_PRIORITY.indexOf(effectiveRole) < ContentConstants.roles.ALL_PRIORITY.indexOf(role);
+                        });
+                        if (hasHigherRole) {
+                            return callback({'code': 401, 'msg': 'You cannot change the role of someone with a higher role than your own'});
+                        }
                     }
```

Comment:
Additionally, can you rename `checkNewContentPermissions` to `checkTargetUserAccess` ? I think it's a bit more representative of what's going on in that method and doesn't imply that it will be looking at the role promotion issues.